### ISSUE TITLE
Global evaluation of interpolation coordinates

### ISSUE THREAD

**Issue by brmather** [underworldcode/underworld2#490]
It would be good if `evaluate_global` broadcasts the interpolated values to all processors. Currently only the root processor returns values. This becomes a problem if a point lies outside the model domain because only the root processor raises a RuntimeError.

https://github.com/underworldcode/underworld2/blob/c37cf9b085a6ef17184ca454c877593629225d2b/underworld/function/_function.py#L752-L754

Not sure if it's feasible here, but returning interpolated values on each processor and a flag, e.g. 0 for interpolation, 1 for extrapolation would be useful.

Happy to help, but I got scared off by all the `comm.send` and `comm.recv` calls.

**Comment by jmansour** [comment]
Broadcasting results to all procs is doable.  But why is it an issue if only the root proc raises the runtime? 

Can you spell out your use case a bit more?

**Comment by brmather** [comment]
Hey John!
If only proc 0 gets a runtime error then the other procs keep chugging away like nothing's happened, which leads to ambiguous errors further down the line. To combat this I have to nest everything within a `try` statement, which is ugly and gross.

**Comment by jmansour** [comment]
Exceptions in parallel are difficult unfortunately. Currently this is our implementation:

https://github.com/underworldcode/underworld2/blob/master/underworld/__init__.py#L187

There are two scenarios:  
1. A subset of procs raise exceptions.
2. All procs raise the same exception. 

In the first scenario, as per the implementation above, the raising procs hit a barrier and wait for all procs to arrive at the same barrier (signifying that all procs have raised).  However, the non-raising procs will never get there, so after a nominal amount of time (15 seconds), the raising procs give up, each one spitting out their error message, and eventually calling `MPI_Abort()` which should tear all others down too.  Granted, this final Mortal Kombat style fatality will catch other procs in random states of operation.

In the second scenario, all procs should reach the same barrier, and then only the root proc will spit out a message.

The idea of all this is that _usually_ users will encounter exceptions collectively, and in these cases we'd rather not have 10000 procs spitting out the same message.  It's far from perfect, but it's generally a reasonable compromise. 

Note that you can bypass all this by setting the `UW_ALL_MESSAGES` environment flag which then reverts to the old behaviour where any raising proc immediately messages and calls `MPI_Abort()`. 

tl;dr:  Try setting the `UW_ALL_MESSAGES` environment flag and see if this gives you more favourable behaviour.

**Comment by jmansour** [comment]
I'll close this for now due to inactivity, but feel free to reopen if needed!
